### ISSUE TITLE
high memory usage and slow performance in large clusters

### ISSUE THREAD

**Issue by juliantaylor** [kubernetes/dashboard#5197]
```
Installation method: helm chart
Kubernetes version: 1.17.3
Dashboard version: 2.01
Metrics provider: metrics-server 0.3.6
Operating system: flatcar linux
```

##### Steps to reproduce
In large clusters with more than 1000 pods the dashboard requires excessive amount of memory and is very slow when looking at namespaces with many pods or looking at all namespaces.

##### Observed result
In a 7000 pod cluster loading the deployment overview page for all namespaces requires more than 24Gi of memory in the dashboard and puts a high load on the metrics which can impact autoscaling workloads as the apiserver fails to answer all requests.

The metrics scraper consumes 5Gi of memory until the dashboard was killed due to OOM.

Only looking at a namespace with 700 pods in it also consumes more than 8Gi memory and the page load takes 30 minutes.

##### Expected result
It would be nice if there was some form of pagination for the fetched resources so looking at very large namespaces or all namespaces the dashboard does dos the apiserver and also does not die due to memory exhaustion.

On pages where pagination may not work (like the namespace overview cpu usage graphs) it should abort when it is not feasible to display the necessary data.

##### Comments
If that is out of scope a way to disable the all namespaces view would help for clusters that are configured to allow authenticated users to see resources likes pods and deployments.

Or a way to restrict the dashboard to certain namespaces so one can deploy multiple dashboard instances with limited view rights in multi tenant namespaces.

**Comment by floreks** [comment]
1. There already is backend-side pagination and you can define on the settings page how many elements should be shown on the list.
2. For large clusters you should definitely disable auto-refresh timer (set it to 0) in the settings view.
3. You might want to tune metrics scraper timers also to scrape metrics less often.
4. K8S API server does not offer a way to paginate and filter the results. That's why we have to download basically the whole cluster at once and our backend consumes a lot of memory due to that. You should extend CPU/MEM limits for the dashboard pod. Disabling the auto-refresh is also crucial for such big clusters.
5. You can restrict Dashboard to a namespace if you properly assign the privileges to the user that uses it. Namespace picker has an option to select custom namespace with the built-in input.

**Comment by juliantaylor** [comment]
Thanks for the fast reply, I figured it may not be configured correctly, but I have not found any documentation on how to tune anything in the docs/ folder.
How can one set these settings via the deployment so they do not have to be manually set in the gui?

There is pagination for some resources like the pod listings but for example the Deployments overview page has a graph with the cpu and memory usage of the deployment. If the deployments have hundreds of pods it appears to fetch all these metrics and go OOM.

How does restricting namespaces work exactly?
I have not found a argument/configuration to limit it during deployment.

If the dashboard users credentials are used to list the namespaces they can still see all namespaces so they would get a huge list of namespaces they cannot use?
Or do you add a special role which can only see specific resource names?

But generally the memory seems to be extremely high, more 24Gi for metrics of 7000 pods is a lot, the dashboard was also not running very long so it shouldn't have much timeseries data.

**Comment by floreks** [comment]
> How can one set these settings via the deployment so they do not have to be manually set in the gui?

Edit the config map using `kubectl`
```bash
kubectl -n kubernetes-dashboard describe configmap kubernetes-dashboard-settings
```

> There is pagination for some resources like the pod listings but for example the Deployments overview page has a graph with the cpu and memory usage of the deployment. If the deployments have hundreds of pods it appears to fetch all these metrics and go OOM.

Top-level graphs will always show summary metrics for all resources available on the view, even if list is paginated. The most important thing is to increase memory limits for the dashboard deployment and metrics scraper. 1 core and 1gb ram should be enough if you disable auto resource refresh.

> How does restricting namespaces work exactly?
I have not found a argument/configuration to limit it during deployment.

This functionality relies on K8S RBACs. You have to define privileges for the user that logs in to the Dashboard using token and restrict privileges to a given namespace/resources.
https://kubernetes.io/docs/reference/access-authn-authz/rbac/

> If the dashboard users credentials are used to list the namespaces they can still see all namespaces so they would get a huge list of namespaces they cannot use?
Or do you add a special role which can only see specific resource names?

Restricting to a single namespace would mean revoking privileges to list namespaces and granting privileges only to get specific namespace. This way namespace selector will be empty and user will have to manually provide namespace name. There is an obvious downside here that the user has to know namespace name. Unfortunately, Kubernetes API does not offer any filtering/whitelisting/blacklisting options. If they ever do, we will definitely use that.

> But generally the memory seems to be extremely high, more 24Gi for metrics of 7000 pods is a lot, the dashboard was also not running very long so it shouldn't have had much timeseries data.

This is quite unrealistic even for our recommended yaml deployment. I am not sure what kind of helm chart are you using (we don't have any official ones yet), but it seems like it might not be configured properly for big clusters.

Metrics scraper takes metrics from the metrics server once every minute and only keeps metrics from the last 15 minutes. Those are very simple metrics that contain only cpu and memory usage. I don't see have it could take that much memory.

Can you provide me the output from `kubectl top` for the dashboard, metrics server and metrics scraper deployments/pods?

**Comment by juliantaylor** [comment]
> Edit the config map using kubectl
What do I put in there? Maybe I am looking at the wrong place for documentation? Is it not in the docs folder of this repository?

> Restricting to a single namespace would mean revoking privileges to list namespaces and granting privileges only to get specific namespace. This way namespace selector will be empty and user will have to manually provide namespace name. There is an obvious downside here that the user has to know namespace name. Unfortunately, Kubernetes API does not offer any filtering/whitelisting/blacklisting options. If they ever do, we will definitely use that.

Users may have multiple namespaces so they should be able to list. Would it be an option to add a configuration to the dashboard which configures which namespaces can be seen regardless of rights of the logged in user?


> This is quite unrealistic even for our recommended yaml deployment. I am not sure what kind of helm chart are you using (we don't have any official ones yet), but it seems like it might not be configured properly for big clusters.

We are using this deployment manifest with no significant changes: https://github.com/kubernetes/dashboard/blob/master/aio/deploy/recommended.yaml
basically just a different namespace and higher memory and cpu limits.

> Can you provide me the output from kubectl top for the dashboard, metrics server and metrics scraper deployments/pods?
here a memory usage graph during clicking on the deployment overview for all namespaces in a 7000 pod cluster, the units are GiB:

![dashboard_memory](https://user-images.githubusercontent.com/542663/83161932-5a4d1480-a109-11ea-87ec-69fd6043c292.png)

the high lines which reset are the dashboard container which gets killed once at 8Gi limit and once at 24Gi limit
The green line that peaks at about 6Gi is the metric scraper.

**Comment by floreks** [comment]
> What do I put in there? Maybe I am looking at the wrong place for documentation? Is it not in the docs folder of this repository?

There is no documentation for this as this is for advanced users only and they already know how to edit resources using `kubectl`. This config map contains our settings structure in a JSON format so it is very easy to update manually. Simply use `kubectl edit` and change resource auto refresh parameter to 0. Maybe also items per page to i.e. 10 (this should be default though).

**Comment by juliantaylor** [comment]
On average our dashboard resource consumption is similarly low, the problem is only when a user clicks on the wrong links (e.g. deployments with too many pods) which then kills the dashboard despite its very high limits.

here are the descriptions:

(skip login is allowed as the resources the dashboard role can see are not secret)

```
Name:                   kubernetes-dashboard
Namespace:              kube-system
CreationTimestamp:      Wed, 06 Jun 2018 19:55:44 +0200
Labels:                 k8s-app=kubernetes-dashboard
                        pipeline-managed=supportive-addons
                        version=v2.0.1
Annotations:            deployment.kubernetes.io/revision: 19
Selector:               k8s-app=kubernetes-dashboard
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           k8s-app=kubernetes-dashboard
                    version=v2.0.1
  Service Account:  kubernetes-dashboard
  Containers:
   kubernetes-dashboard:
    Image:      kubernetesui/dashboard:v2.0.1
    Port:       8443/TCP
    Host Port:  0/TCP
    Args:
      --auto-generate-certificates
      --token-ttl=28800
      --namespace=kube-system
      --enable-skip-login
    Limits:
      cpu:                4
      ephemeral-storage:  250Mi
      memory:             24Gi
    Requests:
      cpu:                100m
      ephemeral-storage:  250Mi
      memory:             1Gi
    Liveness:             http-get https://:8443/ delay=30s timeout=30s period=10s #success=1 #failure=3
    Environment:          <none>
    Mounts:
      /certs from kubernetes-dashboard-certs (rw)
      /tmp from tmp-volume (rw)
  Volumes:
   kubernetes-dashboard-certs:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kubernetes-dashboard-certs
    Optional:    false
   tmp-volume:
    Type:               EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:             
    SizeLimit:          <unset>
  Priority Class Name:  cluster-support

$ kubectl -n kube-system describe deployment dashboard-metrics-scraper 

Name:                   dashboard-metrics-scraper
Namespace:              kube-system
CreationTimestamp:      Thu, 06 Feb 2020 16:46:06 +0100
Labels:                 k8s-app=dashboard-metrics-scraper
                        pipeline-managed=supportive-addons
Annotations:            deployment.kubernetes.io/revision: 10
Selector:               k8s-app=dashboard-metrics-scraper
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           k8s-app=dashboard-metrics-scraper
  Service Account:  kubernetes-dashboard
  Containers:
   dashboard-metrics-scraper:
    Image:      kubernetesui/metrics-scraper:v1.0.4
    Port:       8000/TCP
    Host Port:  0/TCP
    Limits:
      cpu:                4
      ephemeral-storage:  250Mi
      memory:             8Gi
    Requests:
      cpu:                100m
      ephemeral-storage:  250Mi
      memory:             250Mi
    Liveness:             http-get http://:8000/ delay=30s timeout=30s period=10s #success=1 #failure=3
    Environment:          <none>
    Mounts:
      /tmp from tmp-volume (rw)
  Volumes:
   tmp-volume:
    Type:               EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:             
    SizeLimit:          <unset>
  Priority Class Name:  cluster-support

**Comment by juliantaylor** [comment]
> There is no documentation for this as this is for advanced users only and they already know how to edit resources using kubectl. This config map contains our settings structure in a JSON format so it is very easy to update manually. Simply use kubectl edit and change resource auto refresh parameter to 0. Maybe also items per page to i.e. 10 (this should be default though).

The configmap is empty in our cluster and I have checked the role, it does allow the deployment to update it. We are using your recommended manifests, is there some configuration missing in there? The configmap is created empty and there is no reference to it despite in the Role.

**Comment by floreks** [comment]
Unless you actually change some global settings on the settings view it will stay empty because default settings will be used and they are stored in the code. If you change anything, then Dashboard will store the updated version inside the config map. Settings are global and will be then used for everyone.

As a potential improvement maybe we could add option to disable `All namespaces` selector. This way admins of a big clusters could avoid fetching everything at once and force users to work on per-namespace basis. WDYT?

**Comment by juliantaylor** [comment]
ah thank you now I have configuration to change, these is the current config which was used until now:

```
  _global: '{"clusterName":"name","itemsPerPage":10,"logsAutoRefreshTimeInterval":5,"resourceAutoRefreshTimeInterval":5,"disableAccessDeniedNotifications":false}'
```

disabling the all namespaces selector would indeed help, most namespaces are fine and the clusters are too large for the all namespaces view to be useful. It being not visible would avoid users accidentally clicking it and and killing the dashboard in particular as the all namespace selection is right at the top.

**Comment by floreks** [comment]
Check usage after you disable resource autorefresh. It should be much better. I suspect that due to cluster size, overview might not even load with default 5s refresh timer as every 5s it will trigger a new request and they will start overlaping. This will trigger multiple request to the API server that will try to fetch the whole cluster. Overview might even work just fine after changing settings.

**Comment by juliantaylor** [comment]
setting the refresh to 0 did fix the issue, the overviews now load in a reasonable time and memory usage didn't go over 2.5GiB! Thanks.

Though that the autorefresh pilles up requests until it OOms can be considered a bug or?

**Comment by floreks** [comment]
@juliantaylor not really. That's why we offer a way to tune some settings. The thing here is that frontend makes a request to our API and then our API makes request to the K8S API. We can control the requests to our API and cancel them when timer triggers a new one, but it will not cancel the pending request to the K8S API. Due to the nature of how we gather data tuning the settings is the only way right now to control this. You can also set the resource auto refresh to the bigger value like 30/45/60 seconds.

The only way to actually fix this is to use WebSockets and watches instead of regular requests, but we do have an issue for that already. Since this is solved I think we can close.

**Comment by juliantaylor** [comment]
If it can't be fixed then this issue can be closed, thanks

**Comment by davidtseng** [comment]
Note: on one of our large clusters (3400 deployments) I still see errors loading "All namespaces" workloads, even with resourceAutoRefreshTimeInterval set to 0.

**Comment by floreks** [comment]
What was the memory usage? Did it crash with OOM again? I am afraid that the only way here to solve this would be to use cache informers and even that might not help as backend still needs to download the whole cluster into memory in order to apply paging/sorting/filtering. It would be great if API would support this OOTB.

**Comment by floreks** [comment]
I have opened a #5320 to remember about this issue. This is definitely a long term task as it requires major refactoring of our backend but might help with this issue.

**Comment by davidtseng** [comment]
Thanks, with 12G given to kubernetes-dashboard it's not OOM killed but instead after about a minute loading workloads that have a lot of resources it goes to this error page:

MSIE and Chrome friendly error page -->

**Comment by floreks** [comment]
This usually means that connection to the API has been lost. Probably container has been restarted.
